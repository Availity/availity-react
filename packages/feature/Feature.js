import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import BlockUi from 'react-block-ui';
import 'react-block-ui/style.css';
import isFeatureEnabled from './isFeatureEnabled';

const Feature = ({ features, loader, whenDisabled, children, negate }) => {
  const [loading, setLoading] = useState(false);
  const [enabled, setEnabled] = useState(null);

  useEffect(() => {
    // @see https://github.com/sindresorhus/eslint-plugin-unicorn/issues/392
    // eslint-disable-next-line unicorn/consistent-function-scoping
    const checkFeatures = async () => {
      if (!loading) setLoading(true);

      const _enabled = await isFeatureEnabled(features);

      setEnabled(_enabled);

      setLoading(false);
    };

    checkFeatures();
  }, [loading, features]);

  if (loading) {
    if (loader) return loader === true ? <BlockUi blocking /> : loader;
    return null;
  }
  const showChildren = enabled ^ negate; // eslint-disable-line no-bitwise
  if (showChildren) {
    return children;
  }

  return whenDisabled;
};

Feature.propTypes = {
  features: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.string,
      ])
    ),
    PropTypes.string,
  ]).isRequired,
  loader: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]),
  whenDisabled: PropTypes.node,
  children: PropTypes.node,
  negate: PropTypes.bool,
};

Feature.defaultProps = {
  whenDisabled: null,
  children: null,
};

export default Feature;
