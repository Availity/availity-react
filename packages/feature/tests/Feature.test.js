import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Feature from '..';

afterEach(cleanup);

describe('Feature', () => {
  test('should render', () => {
    const { container } = render(<Feature features="1234" loader />);

    expect(container).toMatchSnapshot();
  });
  test('should render with single permission', () => {
    const { container } = render(
      <Feature
        features="1234"
        whenDisabled="You do not have permission to see this"
      />
    );

    expect(container).toMatchSnapshot();
  });
  test('should render with array of features', () => {
    const { container } = render(
      <Feature
        features={['1234', '2345', ['3456', '4567']]}
        whenDisabled="You do not have permission to see this"
      />
    );

    expect(container).toMatchSnapshot();
  });
  test('should render negate features', () => {
    const { container } = render(
      <Feature
        features="1234"
        negate
        whenDisabled="You do not have permission to see this"
      >
        You can see this
      </Feature>
    );

    expect(container).toMatchSnapshot();
  });
});
