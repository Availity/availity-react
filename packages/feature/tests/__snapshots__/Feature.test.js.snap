// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature should render 1`] = `
<div>
  <div
    aria-busy="true"
    class="block-ui undefined"
  >
    <div
      tabindex="0"
    />
    <div
      class="block-ui-container"
      tabindex="0"
    >
      <div
        class="block-ui-overlay"
      />
      <div
        class="block-ui-message-container"
      >
        <div
          class="block-ui-message"
        >
          <div
            class="loading-indicator"
          >
            <span
              class="loading-bullet"
            >
              •
            </span>
             
            <span
              class="loading-bullet"
            >
              •
            </span>
             
            <span
              class="loading-bullet"
            >
              •
            </span>
          </div>
        </div>
      </div>
    </div>
    <span />
  </div>
</div>
`;

exports[`Feature should render negate features 1`] = `<div />`;

exports[`Feature should render with array of features 1`] = `<div />`;

exports[`Feature should render with single permission 1`] = `<div />`;
